### PR TITLE
Relaunch LauncherActivity with NEW_TASK if missing.

### PR DESCRIPTION
**Pull request by PEConn**
I can't quite place how ugly this solution is, but it seems to work...

I've created https://github.com/GoogleChrome/android-browser-helper/issues/27 as well.

### CODE REVIEW

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 1]
```diff
@@ -108,6 +109,28 @@
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
+        boolean hasNewTask = (getIntent().getFlags() & Intent.FLAG_ACTIVITY_NEW_TASK) != 0;
+        boolean hasNewDocument = (getIntent().getFlags() & Intent.FLAG_ACTIVITY_NEW_DOCUMENT) != 0;
+
+        if (!hasNewTask || hasNewDocument) {
+            // If we're created as part of somebody else's Task, relaunch ourselves with NEW_TASK.
```
Comment (beverloo) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: Can we document here _why_ this is important?

**Comment by PEConn** [comment]
@pshmakov Could you check my explanation of the Task gymnastics we do?

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 2]
```diff
@@ -108,6 +109,42 @@
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
+        boolean hasNewTask = (getIntent().getFlags() & Intent.FLAG_ACTIVITY_NEW_TASK) != 0;
+        boolean hasNewDocument = (getIntent().getFlags() & Intent.FLAG_ACTIVITY_NEW_DOCUMENT) != 0;
+
+        if (!hasNewTask || hasNewDocument) {
+            // We only want one instance of the TWA running in the user's Recent Apps. This
```
Comment (pshmakov) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: This comment seems to explain how we ensure it's in a separate task, but not why.

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 3]
```diff
@@ -108,6 +109,42 @@
...
+            // launchMode=singleTask (or singleInstance), however since the TWA consists of a
+            // LauncherActivity which then starts a browser Activity, things get a bit more
+            // complicated.
+
+            // If we used singleTask on LauncherActivity then whenever a TWA was running and a new
+            // Intent was fired, the browser Activity on top would get clobbered.
+
+            // Therefore, we always ensure that LauncherActivity is launched with New Task. This
+            // means that if the TWA is already running a *new* LauncherActivity will be created on
+            // top of the Browser Activity. The browser then launches an Intent with CLEAR_TOP to
```
Comment (pshmakov) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: It seems this code would force this new instance of LauncherActivity to restart as well.
Off the top of my head, I can't think of a way to distinguish LauncherActivity under TWA, mistakenly launched on top of something else, or a LauncherActivity on top of TWA that handles a new intent.

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 4]
```diff
@@ -108,6 +109,42 @@
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
+        boolean hasNewTask = (getIntent().getFlags() & Intent.FLAG_ACTIVITY_NEW_TASK) != 0;
```
Comment (pshmakov) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: Could you extract this whole block into "private boolean restartInNewTask()"?

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java (hunk 3 above)
Comment (PEConn) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: Discussed offline.

**Comment by PEConn** [comment]
Thanks for the review Pavel, all done.

**Pull request merged**